sed/loader/flash/loader.py

```diff
@@ -0,0 +1,615 @@
+        """Resets the index per pulse and electron"""
+        self.index_per_electron = None
+        self.index_per_pulse = None
+
+    def create_multi_index_per_electron(self, h5_file: h5py.File) -> None:
+        """Creates an index per electron using pulseId
+        for usage with the electron resolved pandas dataframe"""
+
+        # Macrobunch IDs obtained from the pulseId channel
+        [train_id, np_array] = self.create_numpy_array_per_channel(
+            h5_file,
+            "pulseId",
+        )
+
+        # Create a series with the macrobunches as index and
+        # microbunches as values
+        macrobunches = (
+            Series(
+                (np_array[i] for i in train_id.index),
+                name="pulseId",
+                index=train_id,
+            )
+            - self._config["dataframe"]["ubid_offset"]
+        )
+
+        # Explode dataframe to get all microbunch vales per macrobunch,
+        # remove NaN values and convert to type int
+        microbunches = macrobunches.explode().dropna().astype(int)
+
+        # Create temporary index values
+        index_temp = MultiIndex.from_arrays(
+            (microbunches.index, microbunches.values),
+            names=["trainId", "pulseId"],
+        )
+
+        # Calculate the electron counts per pulseId
+        # unique preserves the order of appearance
+        electron_counts = index_temp.value_counts()[index_temp.unique()].values
+
+        # Series object for indexing with electrons
+        electrons = (
+            Series(
+                [
+                    np.arange(electron_counts[i])
+                    for i in range(electron_counts.size)
+                ],
+            )
+            .explode()
+            .astype(int)
+        )
+
+        # Create a pandas multiindex using the exploded datasets
+        self.index_per_electron = MultiIndex.from_arrays(
+            (microbunches.index, microbunches.values, electrons),
+            names=["trainId", "pulseId", "electronId"],
+        )
+
+    def create_multi_index_per_pulse(self, train_id, np_array) -> None:
+        """Creates an index per pulse using a pulse resovled channel's
+        macrobunch ID, for usage with the pulse resolved pandas dataframe"""
+
+        # Create a pandas multiindex, useful to compare electron and
+        # pulse resolved dataframes
+        self.index_per_pulse = MultiIndex.from_product(
+            (train_id, np.arange(0, np_array.shape[1])),
+            names=["trainId", "pulseId"],
+        )
+
+    def create_numpy_array_per_channel(
+        self,
+        h5_file: h5py.File,
+        channel: str,
+    ) -> Tuple[Series, np.ndarray]:
+        """Returns a numpy Array for a given channel name for a given file"""
+        # Get the data from the necessary h5 file and channel
+        group = cast(
+            h5py.Group,
+            h5_file[
+                self._config["dataframe"]["channels"][channel]["group_name"]
```

Comment:
with such nested config paths, does it make sense to implement some `/` separated path-like argument to chain these? maybe we can discuss this at some point

Reply:
It is definitely possible but this definitely makes it very readable, don't you think? we can of course use f strings like it's being used in utils.

Reply:
I will resolve this convo now but if you have a suggestion, you could create a issue/PR for it.